node/src/neighborhood/mod.rs

```diff
@@ -507,6 +507,13 @@ impl Neighborhood {
     }
 
     fn handle_route_query_message(&mut self, msg: RouteQueryMessage) -> Option<RouteQueryResponse> {
+        if let Some(ref url) = msg.hostname_opt {
+            if url.contains("0.0.0.0") {
+                error!(self.logger, "Request to wildcard IP detected 0.0.0.0. Most likely because Blockchain Service URL is not set");
```

Comment:
Hmmm, here you prepared a trap for me. Haha. This is a tough one.

Let me try reconstructing what was your thinking. You were going to react on the removal of the null blockchain interface, as you considered that first we had two paths, choosing either for the clandestine or non-clandestine one. Having reduced it to one, you started to think just everything would be routed as clandestine, even the blockchain request-response pairs. And you assumed that it will result in requests eventually heading for a unique route. (A secondary problem of this reasoning is that we probably don't make a new route every time we're handling a new http(s) request. Only if we hit a DNS failure. I might be wrong though.) And so you were thinking that here, in the handler for a route query, you could intercept any blockchain-service happening while the blockchain service url stays unset, because this is default IP you use as a sentinel value in such cases.

The problem is that we've never had such a thing implemented and so we've been blockchain-service non-clandestine ever since. 
In order for it to work, you'd have to build up a link between blockchain interface and the proxy-server.  

I believe this link which you've ignored to which is not there therefore, would be the best place to check for the interfaces readiness to contact a real service. It could actually be a simple flag inside the body of the interface, but preferably, if it could be a new method in the BlockchainInterface trait. That would enforce that no matter what the representation of the interface for a specific chain looks like exactly structurally, the coder implementing a new one will be put facing the request of writing some guts for this new required method in the interface, saying basically only if the service url has been configured yet. 


This is of another issue, but I feel slightly uneasy asking myself to imagine what would be happening if your interface was fully-fledged as it is now and then you only blocked the request from coming out, which would probably cause a couple of time-outs inside our methods written for the scanners. The question is how understandable would the error or warn logs be as we'd better make it really easily told apart from any other and more serious issue to occur. In other words, should we go running down the standard error-handling paths? Or should we cut it sooner and write only specially formed logs that fires off in no other case but the lack of blockchain-service configuration?

Well, even here, some portion of Dan's supervision might be appreciated.

Reply:
**URGENCY STATUS: THIS IS ONE OF THE HIGHEST PRIORITIES AMONG THE URGED FIXES. THE BLOKCHAIN INTERFACE IN YOUR CODE IS UNAWARE THAT IT SHOULD SEND REQUESTS INDIRECTLY THROUGH THE MASQ NETWORK. IT SHOULD BE MADE EXPLICIT THAT IT ALWAYS IS DIRECTED TO THE PROXY SERVER ETC. 

DISCUSSION WITH DAN IS HIGHLY RECOMMENDED**  The problem is that at the moment this might sometimes work as intended and it truly is routed by us to another Node, but probably only if somebody configured his system dns to Node's port. If you do that only for your browser, my understanding is that these blockchain requests will find the shortest way to go towards the blockchain service provider and the MASQ network won't be participating in there.

Reply:
It should be fleshed out to cover up `localhost` and `127.0.0.1`. - Discussion Outcome

Reply:
Hi, I agree that there has been a perfect bite discovered, hopefully already or soon reflected in a card. Well done, whole team. 

The problem is we focused on celebrations and ignored some other aspects of the new design with the BlockchainInterface populated with just quad zeros which are also potentially to cause troubles.

As my initial comments were intended to convey this, I suspect that you falsely assume that the communication between the Node and the blockchain service provider flows through our secure mesh network, which I don't think is true at the moment. It has an implication which you don't expect as I'm afraid of. You thought that these attempts for requests to 0.0.0.0 will be intercepted and halted by your new barrier with the if condition. 
Sadly, the way the interface has been implemented, it stands alone outside our CORES packages etc. 

Therefore I do think it is still possible, that we will be suddenly situated in the middle of the scanning and the Node's effort to contact the assumed external blockchain services won't be blocked.

Then the question is, are you ready for that? Have you thought about how our code would react upon such an incident? 

Maybe this is not decisive...no I think, even if the request were to be routed complexly, it would eventually hit your obstacle put in its way. That's been discussed repeatedly, what hasn't, though, is what will happen with the initiated relation of our interface towards the remote server, because I think, it must be waiting. Of course, it would probably time out, as a behavior of the third party code, but let's rather ask ourselves if we do want to reach so far, causing some unnecessary delays and having to deal with some library specific error? Wouldn't it be much better if we could prevent such calls in their inception state? 

And the funny thing about this is that it would be so convenient to have an "impotent" null interface, yeah, like that one we used to have. Otherwise any different approach seem worse to me... It's not me saying, c'mon, let's revert the work, I'm just tossing ideas around. No clue, at the moment, regarding how to enhance your solution so that it becomes elegant. It's late. I need to stop right here. Brainstorming welcomed.

Reply:
We and the team were under the impression that blockchain calls are only done using the MASQ network under Clandestine mode, but the test reflect that its done in both modes.
Thats why this code was introduced.

Reply:
It's hard to believe for me that this test `debtors_are_credited_once_but_not_twice` has anything to do with this.
Honestly, I don't know of any test which would ever signaled testing a true clandestine mode for the `BlockchainInterface`.

I remember some history and would look the following way literally for years. There would be days when we had two classes defined for the `BlockchainInterface` trait. It would be first the `BlockchainInterfaceNonClandestine` and the other, `BlockchainInterfaceClandestine`. MASQ inherited this from Substratum, where they never got around to implementing the clandestine version. Only one was actually doing something. 

As time was passing on, we didn't care so much about the incompleteness in this pair, such a card has never become a priority. 

Now I'm going to tell you maybe a mildly surprising thing. Do you remember, when you started the works on this card, you dealt with two interfaces there, where one was called `BlockchainInterfaceNull`. This guy was actually a transformed relict of the old times, because I made it directly out of `BlockchainInterfaceClandestine`. Their implementations differed only barely. 

Resulting from the above, the original procedure of initiating the interface was designed as a check of the inputtable parameter `blockchain_service_url`, if the url was there, we went on by the `NonClandestine` mode, otherwise we simply constructed the `BlockchainInterfaceClandestine`. If you still are wondering how that interface was actually functioning, then hear me here that was completely dull, just logging some message along these lines "Blockchain service requests yet cannot be routed clandestinely" and that's also why it later was so tempting to rename it to BlockchainInterfaceNull, because, matter-of-factly, it had been a null class all the time. 

So the previous story is about one way how you can deduce that we cannot make hold of clandestine blockchain services at the moment.

Another approach would be to actually think about the technical page of the current code. I'm pretty sure, for example, there is no communication channel between the BlockchainBridge and the ProxyServer, where in 99% of time this would be the gateway if you want to route something using the MASQ network. (I think Gossip is an exception, but a logical one. It's not even a CORES package actually, and it's obvious that we wouldn't need to know a route, as we send information just to the next-door neighbor. However, the blockchain service requests would definitely have to be wrapped into CORES packages as well as the responses.)
Then you could be thinking if there isn't a way to make a direct order towards the BlockchainServer, then maybe what if it works indirectly? What if we send a request but now besides the browser, what if the request is routed by the same DNS server, as the browser has been pointed to, and therefore we eventually also wind up in the ProxyServer, but as if another request from the browser. 
I don't have a convicing answer for this one. My reasoning is that if the user redirected the requests only from the browser to the Node, then it shouldn't work and the requests travel as vanilla https conversation. 
However, if somebody was so "thorough" and configured the whole system's DNS settings, then I guess, every piece of attempt to contact something outside your LAN would go through MASQ. Updates,... And of course also what we generate by the BlockchainInterface. In other words, It's my belief that the web3 library uses the system's DNS.

I'll probably won't have time to engage in this anymore for I'm leaving tomorrow, but Dan will help you navigate through this. Just let him read my assumptions. 
I make it out even without much thinking that it's rather counter-productive trying to fix into the clandestine version in this card. The take-away should however be that this whole discussion has been unwinding mainly because you've decided to rip out the `BlockchainInterfaceNull` in this card and keep using only the actually transmitting one, using the wild card address 0.0.0.0. That last part might actually be your problem. 

My concern is that you seem not to have full understanding over the changes you're introducing, but more specifically, your measures you deployed to catch any requests towards 0.0.0.0 seems pointless in this different light. 
The requests will actually be executed and I'm uncertain how it will look, like what kind of errors we will be getting? 
I think, if you want to follow this started path, you must test it first on your own to see what you will see in the logs in case that you started the Node without the `blockchain-service-provider-url` parameter set, after which you should use the `scan` command. Make sure nothing eyebrow-raising has occurred.     

At any rate, it might be a good exercise if you confront Dan with this.